specs/DesignDiscussions/DynamicTypeAhead.md

```diff
@@ -0,0 +1,281 @@
+Today with multi select is true the SDK considers the style as Expanded by default. On the same lines when choices.data is valid we will consider style as filtered by default.
+Hence, recommendation is to go with Option 1 and give precedence to choices.data over style.
+
+![img](assets/TypeAhead/Input.ChoiceSet_rendering_2.png)
+
+1.  If choices.data is present and valid then render drop down menu with dynamic choices
+2.  Else check if isMultiSelect is true then render check boxes
+    - Else check style
+      - If style is ChoiceInputStyle.Expanded then render radio buttons
+      - If style is ChoiceInputStyle.Compact then render drop down menu (combo box)
+      - If style is ChoiceInputStyle.Filtered then render filtered drop down menu with static choices
+
+### Fallback impact / Backward compatibility
+
+Dynamic type ahead
+Since we still support defining static values as choices, we can just provide those options and safely ignore the dynamic information. This will make the choice sets a little less functional, but won’t break the host.
+
+### Communication with Host to fetch Dynamic choices
+
+![img](assets/TypeAhead/TypeAheadSearch.png)
+
+1.  On any text change in the type ahead control text box SDK will fetch the dynamic choices from the host.
+2.  SDK will make an asynchronous call on a background thread to the host in order to fetch the dynamic choices.
+3.  The host will resolve the request at its end or by making a call to a backend / bot.
+4.  The host will return a dynamic choices list as response to the SDK.
+5.  The SDK on receiving the dynamic choices will update the choices list (drop down list) of the given type ahead control on the UI thread.
+
+### Debounce Logic
+
+- We will have a default debounce time of `x` ms (e.g 250 ms or 500 ms). After the user has made text change, we add a delay of `x` ms before making a call to the host to fetch the dynamic choices. Host can specify the value in host config. <br/>
+  Recommendation - 250 ms.
+
+### User Experience
+
+We will enable the host to provide it's own UI components and the functionality will be on the SDK side. Host will not need to handle any events which happen on the dynamic ChoiceSet control.
+
+#### _Loading experience_
+
+While the host resolves the request for dynamic choices requested by the sdk we continue to show the static choices and show an indication to the user that dynamic options are being fetched.
+
+- Recommendation - We can add a loader to indicate that dynamic choices are being fetched.
+
+#### _Failure experience_
+
+In case the host is not able to resolve (bot error, network error etc.) the search query from the SDK. Host can respond back with an error message. In this case we will show an error message to indicate failure to fetch dynamic choices.
+Also, we show an error message if the choices are not fetched in fixed time limit (upper limit). This limit can have a default value `x` seconds and also be added to host config.
+
```

Comment:
Can we allow the host to have their own UIs for loading and failure experiences?

Reply:
Yes, we will allow host to provide their own UI for those scenarios. We shall discuss it more during our platform specific design discussions/documents.